test/controllers/works_controller_test.rb

```diff
@@ -0,0 +1,243 @@
+      get edit_work_path(@work)
+      must_respond_with :ok
+    end
+
+    it "responds with NOT FOUND for a fake work" do
+      work_id = Work.last.id + 1
+      get edit_work_path(work_id)
+      must_respond_with :not_found
+    end
+  end
+
+  describe "update" do
+    let (:work_data) {
+      {
+        work: {
+          title: "Updated Title",
+        },
+      }
+    }
+
+    it "changes the data on the work given valid info" do
+      @work.assign_attributes(work_data[:work])
+      expect(@work).must_be :valid?
+      @work.reload
+
+      patch work_path(@work), params: work_data
+
+      must_respond_with :redirect
+      must_redirect_to work_path(@work)
+
+      check_flash
+
+      @work.reload
+      expect(@work.title).must_equal(work_data[:work][:title])
+    end
+
+    it "responds with NOT FOUND for a fake work" do
+      work_id = Work.last.id + 1
+      patch work_path(work_id), params: work_data
+      must_respond_with :not_found
+    end
+
+    it "responds with BAD REQUEST for bad data" do
+      work_data[:work][:title] = ""
+
+      # Assumptions
+      @work.assign_attributes(work_data[:work])
+      expect(@work).wont_be :valid?
+      @work.reload
+
+      # Act
+      patch work_path(@work), params: work_data
+
+      # Assert
+      must_respond_with :bad_request
+
+      check_flash(:error)
+    end
+  end
+
+  describe "destroy" do
+    it "removes the books from the database" do
+      expect {
+        delete work_path(@work)
+      }.must_change "Work.count", -1
+
+      must_respond_with :redirect
+      must_redirect_to works_path
+
+      check_flash
+
+      after_work = Work.find_by(id: @work.id)
+      expect(after_work).must_be_nil
+    end
+
+    it "returns a 404 if the book does not exist" do
+      work_id = -1
+
+      expect(Work.find_by(id: work_id)).must_be_nil
+
+      expect {
+        delete work_path(work_id)
+      }.wont_change "Work.count"
+
+      must_respond_with :not_found
+    end
+  end
+
+  describe "upvote" do
+    before do
+      @user = User.create(username: "generic_name")
+
+      @login_data = {
+        user: {
+          username: @user.username,
+        },
+      }
+    end
+
+    it "flashes an error if user is not logged in" do
+      post logout_path
+      expect(session[:user_id]).must_be_nil
+
+      post upvote_path(@work)
+
+      check_flash(expected_status = :error)
+
+      must_respond_with :redirect
+      must_redirect_to work_path(@work)
+    end
+
+    it "adds the user to works if new vote" do
+      post login_path, params: @login_data
+      expect(session[:user_id]).must_equal @user.id
+
+      expect {
+        post upvote_path(@work)
+      }.must_change "@work.votes.count", +1
+
+      check_flash
+
+      must_respond_with :redirect
+      must_redirect_to work_path(@work)
+    end
+
+    it "flashes an error if user already voted" do
+      post login_path, params: @login_data
+      expect(session[:user_id]).must_equal @user.id
```

Comment:
Good work getting the three interesting cases for `upvote`, especially since the testing logic is so tricky.